### PR TITLE
Adds Data initializer, export as Data, swiftlint and starts error handling

### PR DESCRIPTION
**Pull request by markuswntr**
Allows SwiftGD.Image to be initialized from "raw" data (e.g. Body-Data in a network request)

### CODE REVIEW

**Comment by twostraws** [comment]
I'm loving these changes, Markus – thank you! Are you going to add more to this PR, or should I check and merge as-is? Are any of these breaking changes that require me to bump the major version?

**Comment by markuswntr** [comment]
Hey Paul! Thanks for the reply.
Those are the only changes I have tested so far and should not require a (major) version bump (I just reverted a possible breaking change). There are some more changes I'd like to send you as a PR at a later time as they are not yet tested and definitely not "polished". I'd guess those require a version bump as well, so feel free to review and merge this PR as-is :)

File: Sources/SwiftGD/SwiftGD.swift [hunk 1]
```diff
@@ -49,6 +49,26 @@ public class Image {
...
+
+        // Bytes must not exceed int32 as limit by `gdImageCreate..()`
+        guard data.count < Int32.max else { return nil }
+
+        // Creates a gdImage pointer if given data represents an image in either PNG or JPEG raster format
+        let createImage: (UnsafeMutablePointer<UInt8>) -> gdImagePtr? = { pointer in
+            let size = Int32(data.count)
+            let rawPointer = UnsafeMutableRawPointer(pointer)
+            return gdImageCreateFromPngPtr(size, rawPointer) // Returns if valid PNG image data
+                ?? gdImageCreateFromJpegPtr(size, rawPointer) // Returns if valid JPEG image data
```
Comment (twostraws) [inline Sources/SwiftGD/SwiftGD.swift]: Is `gdImageCreateFromGifPtr()` also available? We might as well go the whole hog.

File: Sources/SwiftGD/SwiftGD.swift [hunk 2]
```diff
@@ -49,6 +49,26 @@ public class Image {
 		}
 	}
 
+    public convenience init?(data: Data) {
+
+        // Bytes must not exceed int32 as limit by `gdImageCreate..()`
+        guard data.count < Int32.max else { return nil }
+
+        // Creates a gdImage pointer if given data represents an image in either PNG or JPEG raster format
```
Comment (twostraws) [inline Sources/SwiftGD/SwiftGD.swift]: "PNG or JPEG" – or WEBP, or potentially GIF as well.

File: Sources/SwiftGD/SwiftGD.swift [hunk 3]
```diff
@@ -299,3 +302,90 @@ public class Color {
...
 }
+
+// MARK: Hexadecimal
+
+extension Color {
+
+    /// The maximum representable integer for each color component.
+    private static let maxHex: Int = 0xff
+
+    /// Initializes a new `Color` instance of given hexadecimal color string.
```
Comment (twostraws) [inline Sources/SwiftGD/SwiftGD.swift]: Nit: "instance of given" should be "instance from a given".

File: Sources/SwiftGD/SwiftGD.swift [hunk 4]
```diff
@@ -299,3 +302,90 @@ public class Color {
...
+    ///
+    /// Given string will be stripped from a single leading "#", if applicable.
+    /// Resulting string must met any of the following criteria:
```
Comment (twostraws) [inline Sources/SwiftGD/SwiftGD.swift]: Nit: "must met" should be "must meet".

File: Sources/SwiftGD/SwiftGD.swift [hunk 5]
```diff
@@ -299,3 +302,90 @@ public class Color {
...
+    /// - Throws: `.invalidColor` if given string does not match any of the above mentioned critera or is not a hex valid color.
+    public convenience init(hex string: String, leadingAlpha: Bool = false) throws {
+        let string = try Color.sanitize(hex: string, leadingAlpha: leadingAlpha)
+        guard let code = Int(string, radix: 16) else {
+            throw Error.invalidColor(reason: "0x\(string) is not a valid hex color code")
+        }
+        self.init(hex: code, leadingAlpha: leadingAlpha)
+    }
+
+    /// Initializes a new `Color` instance of given hexadecimal color values.
```
Comment (twostraws) [inline Sources/SwiftGD/SwiftGD.swift]: Nit: "of given" should be "from given".

File: Sources/SwiftGD/SwiftGD.swift [hunk 6]
```diff
@@ -299,3 +302,90 @@ public class Color {
...
+    ///
+    /// - Parameters:
+    ///   - color: The hexadecimal color value, incl. red, green, blue and alpha
+    ///   - leadingAlpha: Indicate whether given code should be treated as ARGB (`true`) or RGBA (`false`)
+    public convenience init(hex color: Int, leadingAlpha: Bool = false) {
+        let max = Double(Color.maxHex)
+        let first = Double((color >> 24) & Color.maxHex) / max
+        let secnd = Double((color >> 16) & Color.maxHex) / max
```
Comment (twostraws) [inline Sources/SwiftGD/SwiftGD.swift]: There's no need to save characters here – go for `second` and `fourth` for these constants.

File: Sources/SwiftGD/SwiftGD.swift [hunk 7]
```diff
@@ -299,3 +302,90 @@ public class Color {
...
+
+        // Drop leading "#" if applicable
+        var string = string.hasPrefix("#") ? String(string.dropFirst(1)) : string
+
+        // Evaluate if short code w/wo alpha (e.g. `feb` or `feb4`). Double up the characters if so.
+        if string.count == 3 || string.count == 4 {
+            string = string.map({ "\($0)\($0)" }).joined()
+        }
+
+        // Evaluate if fully fledged code w/wo alpha (e.g. `ffaabb` or `ffaabb44`), otherwise throw error
```
Comment (twostraws) [inline Sources/SwiftGD/SwiftGD.swift]: Nit: "if this is a fully fledged code with or without alpha".

File: Sources/SwiftGD/SwiftGD.swift [hunk 8]
```diff
@@ -247,17 +247,22 @@ public class Image {
 }
 
 
-public struct Point {
+public struct Point: Equatable {
```
Comment (twostraws) [inline Sources/SwiftGD/SwiftGD.swift]: 👍

**Comment by twostraws** [comment]
This is turning into a huge PR – that's not necessarily a bad thing, but I am keen to get it merged so we can revise any rough points. Right now my main concerns are:

- Next to nothing from RasterFormat.swift should get out into the public namespace. There's a lot of `public` in there, and although I think Swift locks most of it away I want to be sure that we're exposing as little implementation detail as possible.
- GD only supports raster images, so I think using "Raster" everywhere seems unnecessary. `ExportFormat.jpg` ought to be enough.
- The `export(as:)` method looks great, but I'm not sure of the benefit of `export(using:)` – can you give me an example?
- It looks like loading with the `.any` case tries BMP first, which seems unusual. On the web I would imagine most folks primarily use JPG, PNG, and GIF, then all the rest in no real order.

File: Sources/Color.swift [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    ///   doubled before conversion. Example: `feba` -> `ffeebbaa`
+    /// - Is a string with 3-characters and therefore a shortened hexadecimal color
+    ///   representation **excluding** an alpha component. Given RGB color character will
+    ///   be doubled and an alpha of component of `0xff` (opaque) will be extended before
+    ///   conversion. Example: `feb` -> `ffeebbff`
+    ///
+    /// - Parameters:
+    ///   - string: The hexadecimal color string.
+    ///   - leadingAlpha: Indicate whether given string should be treated as ARGB (`true`) or RGBA (`false`)
+    /// - Throws: `.invalidColor` if given string does not match any of the above mentioned critera or is not a valid hex color.
```
Comment (twostraws) [inline Sources/Color.swift]: Nit: "criteria" :)

File: Sources/SwiftGD/RasterFormat.swift [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+    }
+}
+
+// MARK: - Concrete LibGd RasterFormatter
+
+/// Defines a formatter to be used on BMP import & export conversions
+public struct BMPRasterFormatter: LibGdCompressableRasterFormatter {
+
+    /// The image quality to apply on export operations
+    public var quality: Quality = .default
```
Comment (twostraws) [inline Sources/SwiftGD/RasterFormat.swift]: ? BMP supports compression, but I think it's either "use RLE or don't" – I'm not recall there being any quality aspect.

**Comment by markuswntr** [comment]
Hey Paul, sorry it turned out to become this huge :). I started by simply adding the `export(as: .png)` option you mentioned on Twitter (once almost done, realized that a protocol based solution may suit better) and got wired at some point.

Please find my answers below:

- Next to nothing from RasterFormat.swift should get out into the public namespace. There's a lot of public in there, and although I think Swift locks most of it away I want to be sure that we're exposing as little implementation detail as possible.
> You are totally right.  I will hide everything inside the module excluding `(Importable|Exportable)RasterFormatter` and `(Importable|Exportable)RasterFormat`

- GD only supports raster images, so I think using "Raster" everywhere seems unnecessary. ExportFormat.jpg ought to be enough.
> Will strip `Raster` from related classes

- It looks like loading with the .any case tries BMP first, which seems unusual. On the web I would imagine most folks primarily use JPG, PNG, and GIF, then all the rest in no real order.
> Wupps, My bad! Initially I went with `.jpg, .png, .gif, .webp, .tiff, .bmp, .wbmp` - must have unintentionally changed it while cleaning-up code.

- The export(as:) method looks great, but I'm not sure of the benefit of export(using:) – can you give me an example?
> This would allow others to implement custom exporter/importer directly.
> One example might be a base64 exporter:
> ```
> struct Base64PngExportFormatter: ExportableRasterFormatter {
>
>    func data(of imagePtr: gdImagePtr) throws -> Data {
>        let pngImageData = try ExportableRasterFormat.png.data(of: imagePtr)
>        guard let base64PngImageData = Data(base64Encoded: pngImageData) else {
>            throw Error.invalidImage(reason: "Image is not base64 convertible")
>        }
>        return base64PngImageData
>    }
> }
>
> // Export a base 64 *png* representation
> let image: Image = ...
> let formatter = BASE64ExportFormatter()
> let base64Data = try image.export(using: formatter)
> ```

**Comment by twostraws** [comment]
Thanks for clarifying! Regarding `export(using:)`, that seems to mean we need to expose quite a few internals, and in this case I wonder whether what folks would actually do is use the regular `export(as:)` to get a `Data` containing PNG information (or JPG/GIF/whatever) then convert to base 64 in their own code without needing a custom formatter.

**Comment by markuswntr** [comment]
It would require to expose the `ImportableRasterFormatter`, `ExportableRasterFormatter` and the `RasterFormatter` (which is a typealias to supporting both of the aforementioned formatters).
While typing, I realized that, if we keep the function, the `ExportableRasterFormatter` should actually return a generic value, so you could for example export a base64 *string* (or encrypted image data, image class from another module, ....) if you so desire.

But you are right, all of this can be done without using custom importers/exporters and outside the module.
So it may be overkill, as it is rather convenient than necessary.
I will leave that up to you to decide - it is your repo after all :)

**Comment by twostraws** [comment]
My concern is that `export(using:)` might be *less* convenient than just getting a `Data` and converting to base 64 by hand. Going from `Data` to base 64 is trivial as you've shown, whereas having folks understand `ExportableRasterFormatter` and `gdImagePtr` adds a level of complexity.

So, proposed solution: take out `export(using:)` for now and we can merge the rest. If we get folks asking for this feature, we can look at it again.

Sound good?

**Comment by markuswntr** [comment]
Absolutely :)
Will implement the changes ASAP.
Thanks for the feedback!

**Comment by markuswntr** [comment]
Hey Paul, last push should address remaining issues

**Comment by twostraws** [comment]
This is such a great pull request – brilliant code, and really clearly documented too. Thank you!

**Pull request merged**